Kernel/Devices/FUSEDevice.h

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2024, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Atomic.h>
+#include <Kernel/Devices/CharacterDevice.h>
+#include <Kernel/Locking/Mutex.h>
+#include <Kernel/Locking/SpinlockProtected.h>
+
+namespace Kernel {
+
+struct FUSEInstance {
+    OpenFileDescription const* fd = nullptr;
+    NonnullOwnPtr<KBuffer> pending_request;
+    NonnullOwnPtr<KBuffer> response;
+    bool drop_request = false;
+    bool buffer_ready = false;
+    bool response_ready = false;
+    bool expecting_header = true;
+};
+
+class FUSEDevice final : public CharacterDevice {
+    friend class DeviceManagement;
+
+public:
+    static NonnullLockRefPtr<FUSEDevice> must_create();
+    virtual ~FUSEDevice() override;
+
+    ErrorOr<void> initialize_instance(OpenFileDescription const&);
+    ErrorOr<NonnullOwnPtr<KBuffer>> send_request_and_wait_for_a_reply(OpenFileDescription const&, Bytes);
+    void shutdown_for_description(OpenFileDescription const&);
+
+private:
+    FUSEDevice();
+
+    // ^Device
+    virtual bool is_openable_by_jailed_processes() const override { return false; }
+
+    // ^CharacterDevice
+    virtual ErrorOr<size_t> read(OpenFileDescription&, u64, UserOrKernelBuffer&, size_t) override;
+    virtual ErrorOr<size_t> write(OpenFileDescription&, u64, UserOrKernelBuffer const&, size_t) override;
+    virtual bool can_read(OpenFileDescription const&, u64) const override;
+    virtual bool can_write(OpenFileDescription const&, u64) const override;
+    virtual StringView class_name() const override { return "FUSEDevice"sv; }
+
+    Mutex m_lock { "FUSEDevice"sv };
```

Comment:
Do we still need the `m_lock` locking here? isn't locking `m_instances` enough to synchronize things?